Mathlib/MeasureTheory/Measure/WithDensity.lean

```diff
@@ -379,13 +386,57 @@ theorem set_lintegral_withDensity_eq_set_lintegral_mul_non_measurable₀ (μ : M
   rw [restrict_withDensity hs, lintegral_withDensity_eq_lintegral_mul_non_measurable₀ _ hf h'f]
 #align measure_theory.set_lintegral_with_density_eq_set_lintegral_mul_non_measurable₀ MeasureTheory.set_lintegral_withDensity_eq_set_lintegral_mul_non_measurable₀
 
-theorem withDensity_mul (μ : Measure α) {f g : α → ℝ≥0∞} (hf : Measurable f) (hg : Measurable g) :
+theorem withDensity_mul₀ {μ : Measure α} {f g : α → ℝ≥0∞}
+    (hf : AEMeasurable f μ) (hg : AEMeasurable g μ) :
     μ.withDensity (f * g) = (μ.withDensity f).withDensity g := by
   ext1 s hs
-  simp [withDensity_apply _ hs, restrict_withDensity hs,
-    lintegral_withDensity_eq_lintegral_mul _ hf hg]
+  rw [withDensity_apply _ hs, withDensity_apply _ hs, restrict_withDensity hs,
+    lintegral_withDensity_eq_lintegral_mul₀ hf.restrict hg.restrict]
+
+theorem withDensity_mul (μ : Measure α) {f g : α → ℝ≥0∞} (hf : Measurable f) (hg : Measurable g) :
+    μ.withDensity (f * g) = (μ.withDensity f).withDensity g :=
+  withDensity_mul₀ hf.aemeasurable hg.aemeasurable
 #align measure_theory.with_density_mul MeasureTheory.withDensity_mul
 
+lemma withDensity_inv_same_le {μ : Measure α} {f : α → ℝ≥0∞} (hf : AEMeasurable f μ) :
+    (μ.withDensity f).withDensity f⁻¹ ≤ μ := by
+  change (μ.withDensity f).withDensity (fun x ↦ (f x)⁻¹) ≤ μ
+  rw [← withDensity_mul₀ hf hf.inv]
+  suffices (f * fun x ↦ (f x)⁻¹) ≤ᵐ[μ] 1 by
+    refine (withDensity_mono this).trans ?_
+    rw [withDensity_one]
+  refine ae_of_all _ (fun x ↦ ?_)
+  simp only [Pi.mul_apply, Pi.one_apply]
+  by_cases hx_top : f x = ∞
+  · simp only [hx_top, ENNReal.inv_top, mul_zero, zero_le]
+  by_cases hx_zero : f x = 0
+  · simp only [hx_zero, ENNReal.inv_zero, zero_mul, zero_le]
+  rw [ENNReal.mul_inv_cancel hx_zero hx_top]
+
+lemma withDensity_inv_same₀ {μ : Measure α} {f : α → ℝ≥0∞}
+    (hf : AEMeasurable f μ) (hf_pos : ∀ᵐ x ∂μ, 0 < f x) (hf_ne_top : ∀ᵐ x ∂μ, f x ≠ ∞) :
```

Comment:
Could you assume `∀ᵐ x ∂μ, f x ≠ 0` instead of `hf_pos`? Same thing for the next lemmas.